### PR TITLE
Validation logic fixes

### PR DESCRIPTION
**Pull request by bjosttveit**
## Description

This fixes some problems that were introduced in #2108 relating to validations, and a few regressions that came with #2244. The issues are mainly related to so-called non-incremental validations (`ITaskValidator`s), these validations are not updated on PATCH, and so has to be treated differently since we cannot always assume that they are up to date. The main difficulty is dealing with custom `ITaskValidator`s, since there was previously no way of distinguishing these from custom validations from PATCH. This issue was somewhat sidestepped previously by the fact that the validate data element API was used for initial validation instead of the instance validation API. The former does not run `ITaskValidator`s. This PR adds support for the new `noIncrementalUpdates` field that gets set on validation issues, as well as a workaround for older apps that don't have this feature yet.

**Fixes related to #2108**:
- No longer check validations marked with `noIncrementalUpdates` on submit, rather, check them after process/next fails and we have a fresh set of backend validations. This means you can always retry submit even if these validations are present.
- Don't run non-incremental validations on initial validation or when exiting a subform. For apps with the new backend feature this means adding `?onlyIncrementalUpdates=true`, for older apps it means using the old data element validation API (older apps cannot have multiple datamodels anyway).
- Process/next will run non-incremental validations no matter what, and previously only so-called task-validations (validations without a `field` property) were set. This worked in terms of not blocking subsequent submittion attempts but would potentially not show the reason for why submit failed. For apps with the new backend feature, this is no longer a problem, as these are marked with `noIncrementalUpdates` and treated differently, but for older apps this is a problem. The way I deal with this is by manually setting all backend validations returned from process/next as non-incremental so that we still treat them differently. These should still show up if process/next fails through the `showAllBackendValidations` state.
- The `showAllBackendValidations` state is a sort of fallback mechanism were submit failed, but we cannot necessarily tie the validation issues to visible nodes, and so we just show all of the backend validations to give some feedback to the user. To get into this state, submit has to fail even though we don't have any "normal" validations that we can show the user. So this is deals with both showing `ITaskValidator` validations and the edge case where there are backend validations that cannot be tied to visible nodes. The main changes here is that we now re-try validations including non-incremental validations at certain times, like after save and when instance data changes, to try to remove these validations as they are fixed instead of the user having to try to submit again while errors are still visible.

**Regressions from #2244**:
- Fixes regression with validation visibility, where page validation could override a components `showValidations` property by hiding validations that were supposed to always be visible for that component. Also added unit tests for this.
- Fixes regression with validation visibility, where visbility is not removed once all validations of a specific category are fixed. For example, if `required` validations becomes visible after trying to submit, fixing that validation should return visibility back so that `required` does not show up again until you try to submit again. Added unit tests for this case.

**Other changes**:
- I had some issues with infinite re-rendering which seemed to be caused by the usage of `useTaskStore` returning new objects in the selector without using shallow comparison on the top-level objects, [see zustand docs](https://zustand.docs.pmnd.rs/guides/prevent-rerenders-with-use-shallow). I changed these to use multiple selectors instead selecting a single field at a time.
- I removed the use of `waitForAnimationFrames` which was used for trying to wait until validation state was synced to nodes, and instead added a more explicit method of checking the last processed validations similar to how we do it in the validation context instead. The motivation here was mainly that `requestAnimationFrame` was causing some problems with jest, and that it was not a very robust method to begin with. I hope the performance implications are not too big as this will cause a few more updates to NodeData.

## Related Issue(s)

- #2108 

## Verification/QA

- Manual functionality testing
  - [x] I have tested these changes manually
  - [ ] Creator of the original issue (or service owner) has been contacted for manual testing (or will be contacted when released in alpha)
  - [ ] No testing done/necessary
- Automated tests
  - [x] Unit test(s) have been added/updated
  - [ ] Cypress E2E test(s) have been added/updated
  - [ ] No automatic tests are needed here (no functional changes/additions)
  - [ ] I want someone to help me make some tests
- UU/WCAG ([follow these guidelines](https://aksel.nav.no/god-praksis/artikler/utvikling?tema=universell-utforming) until we have our own)
  - [ ] I have tested with a screen reader/keyboard navigation/automated wcag validator
  - [x] No testing done/necessary (no DOM/visual changes)
  - [ ] I want someone to help me perform accessibility testing
- User documentation @ [altinn-studio-docs](https://github.com/Altinn/altinn-studio-docs)
  - [ ] Has been added/updated
  
  - [x] No functionality has been changed/added, so no documentation is needed
  - [ ] I will do that later/have created an issue
  
- Support in Altinn Studio
  - [ ] Issue(s) created for support in Studio
  
  - [x] This change/feature does not require any changes to Altinn Studio
- Sprint board
  - [x] The original issue (or this PR itself) has been added to the Team Apps project and to the current sprint board
  - [ ] I don't have permissions to do that, please help me out
- Labels
  - [x] I have added a `kind/*` label to this PR for proper release notes grouping
  - [ ] I don't have permissions to add labels, please help me out

### CODE REVIEW

File: src/features/validation/ValidationStorePlugin.tsx [hunk 1]
```diff
@@ -171,19 +180,18 @@ export class ValidationStorePlugin extends NodeDataPlugin<ValidationStorePluginC
             // constantly recompute this.
             const state = zustand.getState();
 
-            const out: string[] = [];
+            const out_nodes: string[] = [];
+            const out_validations: AnyValidation[] = [];
```
Comment (olemartinorg) [inline src/features/validation/ValidationStorePlugin.tsx:184]: Ooh, snake_case! Exotic! :snake: :joy: Sorry, just feel like I didn't have enough comments to match the size of this PR. That's to say, _great work_! :muscle:

**Review by olemartinorg — APPROVED** [review]

File: src/features/validation/backendValidation/backendValidationQuery.ts [hunk 2]
```diff
@@ -1,84 +1,154 @@
...
+  const queryFn = useBackendValidationQueryFunc(onlyIncrementalValidators, forceMarkNoIncrementalUpdates);
+  const queryKey = useBackendValidationQueryKey();
   const client = useQueryClient();
 
-  return useCallback(
-    () => client.invalidateQueries({ queryKey: ['validation', instanceId, currentProcessTaskId, true] }),
-    [client, currentProcessTaskId, instanceId],
-  );
+  return useCallback(() => {
+    // To fetch the query again with a potentially different query function, we first need to invalidate the existing query without triggering an automatic refetch
```
[the remark this answers is not in the document]
Comment (bjosttveit) [reply src/features/validation/backendValidation/backendValidationQuery.ts:88]: I cannot refetch with a different query function (read URL) than the one used in the `useQuery` that returns the refetch function. In my case, I need to be able to refetch with or without `onlyIncrementalValidators`. The function used in `useBackendValidationQuery` never changes, and is really only used for initial validation when loading the app

[the remark this answers is not in the document]
Comment (bjosttveit) [reply src/features/validation/backendValidation/backendValidationQuery.ts:88]: I could have fetched it manually and just updated the query cache directly, but then I cannot know elsewhere if the query `isFetching` or not.

File: src/features/validation/ValidationStorePlugin.tsx (hunk 1 above)
Comment (bjosttveit) [reply src/features/validation/ValidationStorePlugin.tsx:184]: I guess I have been using too much Python in my spare time, I get confused 🤣

**Review by danielskovli — APPROVED** [review]
Some of this is a bit over my head, but otherwise looks good 🎉

**Pull request merged**